plugins/cryptolib/cryptolib.cpp

```diff
@@ -0,0 +1,705 @@
+
+        if(!SHA512_Update(&context, (unsigned char*)inputData, lenInputData))
+            throwHashError("OpenSSL ERROR calling SHA512_Update");
+
+        __lenResult = SHA512_DIGEST_LENGTH;//64 bytes
+        __result = CTXMALLOC(parentCtx, __lenResult);
+
+        if(!SHA512_Final((unsigned char *)__result, &context))
+        {
+            free(__result);
+            throwHashError("OpenSSL ERROR calling SHA512_Final");
+        }
+    }
+    else
+    {
+        throw MakeStringException(-1, "Unsupported hash algorithm '%s' specified", algorithm);
+    }
+}
+
+//-------------------------------------------------------------------------------------------------------------------------------------------
+// PUBLIC KEY SUPPORT
+//-------------------------------------------------------------------------------------------------------------------------------------------
+
+CRYPTOLIB_API void CRYPTOLIB_CALL clInstalledPublicKeyAlgorithms(bool & __isAllResult, size32_t & __lenResult, void * & __result)
+{
+    __isAllResult = false;
+    StringArray algorithms;
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+    CriticalBlock block(loadingDigestCS);
+    if (!digestsLoaded)
+    {
+        OpenSSL_add_all_digests();
+        digestsLoaded = true;
+    }
+#endif
+    OBJ_NAME_do_all_sorted(OBJ_NAME_TYPE_PKEY_METH, my_clAlgorithmsCallback, (void*)&algorithms);
+    stringArrayToECLSetOfString(algorithms, &__result, &__lenResult);
+}
+
+CRYPTOLIB_API void CRYPTOLIB_CALL clSupportedPublicKeyAlgorithms(bool & __isAllResult, size32_t & __lenResult, void * & __result)
+{
+    __isAllResult = false;
+    StringArray algorithms;
+    algorithms.appendList("SHA256", ",");
+    stringArrayToECLSetOfString(algorithms, &__result, &__lenResult);
+}
+
+//signature helper function
+void doPKISign(size32_t & __lenResult, void * & __result,
+        IDigitalSignatureManager * pDSM,
+        size32_t lenInputdata, const void * inputdata)
+{
+    StringBuffer sbSig;
+    if (pDSM->digiSign(sbSig, lenInputdata, inputdata))
+    {
+        __result = CTXMALLOC(parentCtx, sbSig.length());
+        __lenResult = sbSig.length();
+        memcpy(__result, sbSig.str(), __lenResult);
+    }
+    else
+    {
+        throw MakeStringException(-1, "Unable to create Digital Signature");
+    }
+}
+
+void verifyPKIAlgorithm(const char * pkAlgorithm)
+{
+    if (0 != stricmp(pkAlgorithm, "SHA256"))
+        throw MakeStringException(-1, "Unsupported PKI algorithm (%s) specified", pkAlgorithm);
+}
+
+
+//-----------------------------------------------------------------
+//  Simple cache for instances of Digital Signature Managers
+//-----------------------------------------------------------------
+class CDSMCache
+{
+private:
+    CriticalSection signingDSMCacheCS;//guards modifications to the cache map
+    CriticalSection verifyingDSMCacheCS;//guards modifications to the cache map
+    typedef std::map<string, IDigitalSignatureManager *> DSMCache;
```

Comment:
The digital signature managers will never be freed.  Is that a intentional or a problem?  It should be noted, or use Owned<IDSM>